### ISSUE TITLE
parallel

### ISSUE THREAD

**Issue by congmonkey** [bob48523/modified_wrn#6]
# -*- coding: utf-8 -*-

# License: BSD
# Author: Sasank Chilamkurthy

from __future__ import print_function, division

import argparse

from models import *
import config_parallel as cf
from preprocess import *

import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable
import torch.backends.cudnn as cudnn

import torchvision
from torchvision import datasets, models, transforms

import matplotlib.pyplot as plt
import time
import copy
import os
import setproctitle
import numpy as np
import random

os.environ["CUDA_VISIBLE_DEVICES"] = "4"

parser = argparse.ArgumentParser(description='PyTorch VGG Training')
parser.add_argument('--resume', '-r', action='store_true', help='resume from checkpoint')
parser.add_argument('--save')
parser.add_argument('--testOnly', '-t', action='store_true', help='Test mode with the saved model')
parser.add_argument('--check', '-c', action='store_true', help='Check the saved model')

args = parser.parse_args()
args.save = args.save or './parallelb.base'
setproctitle.setproctitle(args.save)

use_cuda = torch.cuda.is_available()
best_acc = 0  # best test accuracy
start_epoch = cf.start_epoch  # start from epoch 0 or last checkpoint epoch
epochs = cf.num_epochs
######################################################################
# Load Data

data_dir = {
    'cifar': '/ssd/RookieProject/CIFAR100/',
    'tiny': '/ssd/RookieProject/TinyImageNet',
    'vgg': '/ssd/RookieProject/VGGFlowers',
}

test_transforms = {
    x: transforms.Compose([
        transforms.RandomSizedCrop(64),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize(cf.mean[x], cf.std[x])
    ])
    for x in ['cifar', 'tiny', 'vgg']
}

val_transforms = {
    x: transforms.Compose([
        transforms.Scale(72),
        transforms.CenterCrop(64),
        transforms.ToTensor(),
        transforms.Normalize(cf.mean[x], cf.std[x])
    ])
    for x in ['cifar', 'tiny', 'vgg']
}

trainset = {
    'cifar': TxtFolder(os.path.join(data_dir['cifar']), 'list_train.txt', test_transforms['cifar']),
    'tiny': datasets.ImageFolder(os.path.join(data_dir['tiny'], 'train'), test_transforms['tiny']),
    'vgg': datasets.ImageFolder(os.path.join(data_dir['vgg'], 'train'), test_transforms['vgg']),
}

valset = {
    'cifar': TxtFolder(os.path.join(data_dir['cifar']), 'list_val.txt', val_transforms['cifar']),
    'tiny': datasets.ImageFolder(os.path.join(data_dir['tiny'], 'val'), val_transforms['tiny']),
    'vgg': datasets.ImageFolder(os.path.join(data_dir['vgg'], 'val'), val_transforms['vgg']),
}

trainloader = {
    x: torch.utils.data.DataLoader(trainset[x], batch_size=cf.batch_size[x],
                                   shuffle=True, num_workers=4)
    for x in ['cifar', 'tiny', 'vgg']
}

valloader = {
    x: torch.utils.data.DataLoader(valset[x], batch_size=cf.batch_size[x],
                                   shuffle=False, num_workers=4)
    for x in ['cifar', 'tiny', 'vgg']
}

use_gpu = torch.cuda.is_available()

class AverageMeter(object):
    """Computes and stores the average and current value"""
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count
        
if (args.check):
    print('\n[Check Phase] : Model setup')
    checkpoint = torch.load(os.path.join(args.save, 'cifar.t7'))
    net = checkpoint['net']
    if use_gpu:
        model_ft = net.cuda()
        model_ft = torch.nn.DataParallel(net, device_ids=[0])
        cudnn.benchmark = True  
    CheckF = open(os.path.join(args.save, 'weight.txt'), 'w')
    CheckF.write('{}\n'.format(net))
    
    for m in net.modules():
        if isinstance(m, nn.Conv2d):
            CheckF.write('{}\n{}\n'.format(m, m.weight.data))
        elif isinstance(m, nn.Linear):
            CheckF.write('{}\n{}\n'.format(m, m.weight.data))

    sys.exit(0)
    
######################################################################
# Training the model
if (args.testOnly):
    print('\n[Test Phase] : Model setup')
    #assert os.path.isdir('checkpoint'), 'Error: No checkpoint directory found!'
    checkpoint = torch.load(os.path.join(args.save, 'cifar.t7'))
    net = checkpoint['net']

    classes = {'cifar': 100, 
               'tiny': 1000, 
               'vgg': 102}
    if use_cuda:
        net.cuda()
       # net = torch.nn.DataParallel(net, device_ids=range(torch.cuda.device_count()))
       # cudnn.benchmark = True
    net.eval()
    trainW = {x: open(os.path.join(args.save, 'class_' + x + '.csv'), 'w') for x in ['cifar','tiny','vgg']}
    
    for data_name in ['cifar','tiny','vgg']:
        test_loss = 0
        correct = 0
        total = 0   
        #print (classes[data_name])
        
        #compute_class(valloader[data_name], net, classes[data_name], args.save, data_name)
        class_correct = list(0. for i in range(classes[data_name]))
        class_total = list(0. for i in range(classes[data_name]))        
        for batch_idx, (inputs, targets) in enumerate(valloader[data_name]):
            if use_cuda:
                inputs, targets = inputs.cuda(), targets.cuda()
            target = targets
            inputs, targets = Variable(inputs, volatile=True), Variable(targets)
            cifaro, tinyo, vggo = net(inputs, inputs, inputs)

            if data_name == 'cifar':
                outputs = cifaro
            elif data_name == 'tiny':
                outputs = tinyo
            elif data_name == 'vgg':
                outputs = vggo
                
            _, predicted = torch.max(outputs.data, 1)
            c = (predicted == target).squeeze()
            total += targets.size(0)
            correct += predicted.eq(targets.data).cpu().sum()
            
            for i in range(len(targets)):
                #print (len(targets))
                label = target[i]
#                
                class_correct[label] += c[i]
                class_total[label] += 1
            
        for i in range(classes[data_name]):
            trainW[data_name].write('{},{},{},{}\n'.format(i, class_correct[i], class_total[i], class_correct[i]/class_total[i]))
            trainW[data_name].flush()
        trainW[data_name].close()

    # Save checkpoint when best model
        acc = 100.*correct/total
        print("| Test " + data_name +" Result\tAcc@1: %.2f%%" %(acc))

    sys.exit(0)
    
def train(net, criterion, epoch, trainF):
    print('\nEpoch: %d' % epoch)
    net.train()
    train_loss = 0
    top1 = {x: AverageMeter() for x in ['cifar','tiny','vgg']}
    top5 = {x: AverageMeter() for x in ['cifar','tiny','vgg']}
    losses = {x: AverageMeter() for x in ['cifar','tiny','vgg']}
#    correct = 0
    total = 0
    nProcessed = 0
    k = {
     'cifar': 1,
     'tiny': 0.5,
     'vgg': 0.5,
    }
    
    N= len(trainloader['cifar'])
    #nTrain = len(trainloader[data_name].dataset)
    idx = [i for i in range(N)]
    optimizer = optim.SGD(net.parameters(), lr=cf.adjust_rate(epoch), momentum=cf.momentum, weight_decay=1e-4)
    dataiter = {x: iter(trainloader[x]) for x in ['cifar','tiny','vgg']}
    for batch_idx in idx:
           
        #inputs, targets = {x: dataiter[x].next() for x in ['cifar','tiny','vgg']}
        inputs = {}
        targets = {}
        for x in ['cifar','tiny','vgg']:
            inputs[x], targets[x] = dataiter[x].next()
            if use_cuda:
                inputs[x], targets[x] = inputs[x].cuda(), targets[x].cuda()
#        inputs_cifar, targets_cifar = dataiter['cifar'].next()
#        inputs_tiny, targets_tiny = dataiter['tiny'].next()
#        inputs_vgg, targets_vgg = dataiter['vgg'].next()
#        if use_cuda:
#            inputs_cifar, targets_cifar = inputs_cifar.cuda(), targets_cifar.cuda()
#            inputs_tiny, targets_tiny = inputs_tiny.cuda(), targets_tiny.cuda()
#            inputs_vgg, targets_vgg = inputs_vgg.cuda(), targets_vgg.cuda()
            
        optimizer.zero_grad()
        target = {x: targets[x] for x in ['cifar','tiny','vgg']}
        inputs= {x: Variable(inputs[x]) for x in ['cifar','tiny','vgg']}
        targets= {x: Variable(targets[x]) for x in ['cifar','tiny','vgg']}
#        target_cifar = targets_cifar
#        target_tiny = targets_tiny
#        target_vgg = targets_vgg

        
#        inputs_cifar, targets_cifar = Variable(inputs_cifar), Variable(targets_cifar)
#        inputs_tiny, targets_tiny = Variable(inputs_tiny), Variable(targets_tiny)
#        inputs_vgg, targets_vgg = Variable(inputs_vgg), Variable(targets_vgg)
        
        cifaro, tinyo, vggo = net(inputs['cifar'], inputs['tiny'], inputs['vgg'])
        outputs = {
            'cifar': cifaro, 
            'tiny': tinyo,
            'vgg': vggo,
        }
            
        #make_graph.save('./t.dot', loss.creator); assert(False)
        loss = k['cifar']*criterion(outputs['cifar'], targets['cifar']) +\
        k['tiny']*criterion(outputs['tiny'], targets['tiny']) +\
        k['vgg']*criterion(outputs['vgg'], targets['vgg'])
               
        loss.backward()
        optimizer.step()
        
        for x in ['cifar','tiny','vgg']:
            _, predicted = torch.max(outputs[x].data, 1)
            total += targets[x].size(0)
            prec1, prec5 = accuracy(outputs[x].data, target[x], topk=(1, 5))
            top1[x].update(prec1[0], inputs[x].size(0))
            top5[x].update(prec5[0], inputs[x].size(0))
            losses[x].update(loss.data[0], inputs[x].size(0))
            
#        _, predicted = torch.max(outputs.data, 1)
#        total += targets.size(0)
#        prec1, prec5 = accuracy(outputs.data, target, topk=(1, 5))
#        top1['cifar'].update(prec1[0], inputs.size(0))
#        top5['cifar'].update(prec5[0], inputs.size(0))
#        losses['cifar'].update(loss.data[0], inputs.size(0))
            print('Train Epoch: [{}/{} ({:.0f}%)]\t'.format(batch_idx, N, 100. * batch_idx / N))
            print(x+' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
              .format(top1=top1[x], top5=top5[x]))
        
    for x in ['cifar','tiny','vgg']:    
        trainF[x].write('{},{},{},{}\n'.format(epoch, losses[x].avg, top1[x].avg, top5[x].avg))
        trainF[x].flush()
        
def test(net, epoch, testF):
    global best_acc
    net.eval()
    top1 = {x: AverageMeter() for x in ['cifar','tiny','vgg']}
    top5 = {x: AverageMeter() for x in ['cifar','tiny','vgg']}
    losses = {x: AverageMeter() for x in ['cifar','tiny','vgg']}     
    k = {
     'cifar': 0.33,
     'tiny': 0.33,
     'vgg': (1-0.66),
    }
    err_total = 0
    for data_name in ['cifar','tiny','vgg']:
        test_loss = 0
        incorrect = 0
        correct = 0
        total = 0
        
        for batch_idx, (inputs, targets) in enumerate(valloader[data_name]):
            if use_cuda:
                inputs, targets = inputs.cuda(), targets.cuda()
            target = targets
            inputs, targets = Variable(inputs, volatile=True), Variable(targets)
            cifaro, tinyo, vggo = net(inputs, inputs, inputs)
            
            if data_name == 'cifar':
                loss = criterion(cifaro, targets)
            elif data_name == 'tiny':
                loss = criterion(tinyo, targets)
            elif data_name == 'vgg':
                loss = criterion(vggo, targets) 

            if data_name == 'cifar':
                outputs = cifaro

            elif data_name == 'tiny':
                outputs = tinyo
            
            elif data_name == 'vgg':
                outputs = vggo

            test_loss += loss.data[0]
            _, predicted = torch.max(outputs.data, 1)
            total += targets.size(0)
            correct += predicted.eq(targets.data).cpu().sum()
            incorrect += predicted.ne(targets.data).cpu().sum()
            prec1, prec5 = accuracy(outputs.data, target, topk=(1, 5))
            top1[data_name].update(prec1[0], inputs.size(0))
            top5[data_name].update(prec5[0], inputs.size(0))

        nTotal = len(valloader[data_name].dataset)
        err  = 100.*incorrect/nTotal
        test_loss /= len(valloader[data_name])
        err_total += k[data_name]*err
        
        #print (err_total)
    
        print('\nTest set: Average loss: {:.4f}, Error: {}/{} ({:.0f}%)\n'.format(
            test_loss, incorrect, nTotal, err))
        print(' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
              .format(top1=top1[data_name], top5=top5[data_name]))
        testF[data_name].write('{},{},{},{}\n'. format(epoch, test_loss, top1[data_name].avg, top5[data_name].avg))
        testF[data_name].flush()
        
    return err_total
#    # Save checkpoint.
#    acc = 100.*correct/total
#    
#    state = {
#        'net': net.module if use_cuda else net,
#        #'net': net,
#        'acc': acc,
#        'epoch': epoch,
#    }
#    torch.save(state, os.path.join(args.save, 'cifar.t7'))
#    if acc > best_acc:
#        torch.save(state, os.path.join(args.save, 'cifar.t7'))
#        best_acc = acc 
            

######################################################################
# Visualizing the model predictions
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#
# Generic function to display predictions for a few images
#

def accuracy(output, target, topk=(1,)):
    """Computes the precision@k for the specified values of k"""
    maxk = max(topk)
    batch_size = target.size(0)

    _, pred = output.topk(maxk, 1, True, True)
    pred = pred.t()
    correct = pred.eq(target.view(1, -1).expand_as(pred))

    res = []
    for k in topk:
        correct_k = correct[:k].view(-1).float().sum(0)
        res.append(correct_k.mul_(100.0 / batch_size))
    return res

######################################################################
# Finetuning the convnet
# ----------------------
# Model
if args.resume:
    # Load checkpoint.
    print('==> Resuming from checkpoint..')
    #assert os.path.isdir('checkpoint'), 'Error: no checkpoint directory found!'
    checkpoint = torch.load(os.path.join(args.save, 'cifar.t7'))
    model_ft = checkpoint['net']
    best_acc = checkpoint['acc']
    start_epoch = checkpoint['epoch']
else:
    print('==> Building new model..')
    model_ft = Resnet_sen_share(PreActBlock, [2,2,2,2], 100, 1000, 102)
    #net = ResNet34()
    print ('Number of parames:{}'.format(
        sum([p.data.nelement() for p in model_ft.parameters()])))
    
if use_gpu:
    model_ft = model_ft.cuda()
    model_ft = torch.nn.DataParallel(model_ft, device_ids=[0])
    cudnn.benchmark = True

criterion = nn.CrossEntropyLoss()

######################################################################
# Train and evaluate
# ^^^^^^^^^^^^^^^^^^

if args.resume:
    trainF = {x: open(os.path.join(args.save, 'train_' + x + '.csv'), 'a')
    for x in ['cifar', 'tiny', 'vgg']
    }
    testF = {x: open(os.path.join(args.save, 'test_' + x + '.csv'), 'a')
    for x in ['cifar', 'tiny', 'vgg']
    }
else:
    trainF = {x: open(os.path.join(args.save, 'train_' + x + '.csv'), 'w')
    for x in ['cifar', 'tiny', 'vgg']
    }
    testF = {x: open(os.path.join(args.save, 'test_' + x + '.csv'), 'w')
    for x in ['cifar', 'tiny', 'vgg']
    }  
    
elapsed_time = 0
#epochs = 300
for epoch in range(start_epoch, start_epoch+epochs):
    err = 0    
    start_time = time.time()
    train(model_ft, criterion, epoch, trainF)
    err = test(model_ft, epoch, testF)
    epoch_time = time.time() - start_time
    elapsed_time += epoch_time
    print('| Elapsed time : %d:%02d:%02d'  %(cf.get_hms(elapsed_time)))
    #    # Save checkpoint.
#    acc = 100.*correct/total
   
    state = {
        'net': model_ft.module if use_cuda else model_ft,
        #'net': net,
        'acc': 100-err,
        'epoch': epoch,
    }
    print (100-err)
    print (best_acc)
    torch.save(state, os.path.join(args.save, 'cifar_latest.t7'))
    if (100-err) > best_acc:
        torch.save(state, os.path.join(args.save, 'cifar.t7'))
        best_acc = 100-err 

for x in ['cifar', 'tiny', 'vgg']:  
    trainF[x].close()
    testF[x].close()
